cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -0,0 +1,1722 @@
+  std::chrono::nanoseconds ns_count(static_cast<int64_t>(secs) * 1000000000);
+  return TimePoint(std::chrono::duration_cast<TimePoint::duration>(ns_count));
+}
+
+void FileObjectToInfo(
+    const Azure::Storage::Files::DataLake::Models::PathProperties& properties,
+    FileInfo* info) {
+  info->set_type(FileType::File);
+  info->set_size(static_cast<int64_t>(properties.FileSize));
+  info->set_mtime(ToTimePoint((int)(std::chrono::duration_cast<std::chrono::seconds>(
+                                        properties.LastModified - Azure::DateTime(1970))
+                                        .count())));
+}
+
+void PathInfoToFileInfo(const std::string path, const FileType type, const int64_t size,
+                        const Azure::DateTime dt, FileInfo* info) {
+  info->set_type(type);
+  info->set_size(size);
+  info->set_path(path);
+  info->set_mtime(ToTimePoint(
+      (int)(std::chrono::duration_cast<std::chrono::seconds>(dt - Azure::DateTime(1970))
+                .count())));
+}
+
+}  // namespace
+
+// -----------------------------------------------------------------------
+// Azure filesystem implementation
+
+class AzureBlobFileSystem::Impl
+    : public std::enable_shared_from_this<AzureBlobFileSystem::Impl> {
+ public:
+  io::IOContext io_context_;
+  std::shared_ptr<Azure::Storage::Blobs::BlobServiceClient> gen1Client_;
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakeServiceClient> gen2Client_;
+  std::string dfs_endpoint_url;
+  std::string blob_endpoint_url;
+  bool isHierarchicalNamespaceEnabled;
+
+  explicit Impl(AzureOptions options, io::IOContext io_context)
+      : io_context_(io_context), options_(std::move(options)) {}
+
+  Status Init() {
+    dfs_endpoint_url = options_.account_dfs_url;
+    blob_endpoint_url = options_.account_blob_url;
+    RETURN_NOT_OK(InitServiceClient(gen1Client_, options_, blob_endpoint_url));
+    RETURN_NOT_OK(InitServiceClient(gen2Client_, options_, dfs_endpoint_url));
+    if (options_.isTestEnabled) {
+      // gen1Client_->GetAccountInfo().Value.IsHierarchicalNamespaceEnabled throws error
+      // in azurite
+      isHierarchicalNamespaceEnabled = false;
+    } else {
+      isHierarchicalNamespaceEnabled =
+          gen1Client_->GetAccountInfo().Value.IsHierarchicalNamespaceEnabled;
+    }
+    return Status::OK();
+  }
+
+  const AzureOptions& options() const { return options_; }
+
+  // Create a container. Successful if container already exists.
+  Status CreateContainer(const std::string& container) {
+    auto fileSystemClient = gen2Client_->GetFileSystemClient(container);
+    fileSystemClient.CreateIfNotExists();
+    return Status::OK();
+  }
+
+  // Tests to see if a container exists
+  Result<bool> ContainerExists(const std::string& container) {
+    auto fileSystemClient = gen2Client_->GetFileSystemClient(container);
+    try {
+      auto properties = fileSystemClient.GetProperties();
+      return true;
+    } catch (std::exception const& e) {
```

Comment:
Should we not inspect the exception to find out what went wrong? Perhaps we should really error out in some cases?